src/gateway/Gateway.sol

```diff
@@ -254,26 +254,47 @@ contract Gateway is Auth, IGateway {
 
         uint256 numRouters = routers.length;
         require(numRouters > 0, "Gateway/routers-not-initialized");
-        uint256 fuel = quota;
-        uint256 tank = fuel > 0 ? fuel : gasService.shouldRefuel(source, message) ? address(this).balance : 0;
 
+        uint256 fuel = quota;
         uint256 messageCost = gasService.estimate(message);
         uint256 proofCost = gasService.estimate(proof);
 
-        for (uint256 i; i < numRouters; i++) {
-            RouterLike currentRouter = RouterLike(routers[i]);
-            bool isPrimaryRouter = i == PRIMARY_ROUTER_ID - 1;
-            bytes memory payload = isPrimaryRouter ? message : proof;
+        if (fuel > 0) {
+            uint256 tank = fuel;
+            for (uint256 i; i < numRouters; i++) {
+                RouterLike currentRouter = RouterLike(routers[i]);
+                bool isPrimaryRouter = i == PRIMARY_ROUTER_ID - 1;
+                bytes memory payload = isPrimaryRouter ? message : proof;
 
-            uint256 consumed = currentRouter.estimate(payload, isPrimaryRouter ? messageCost : proofCost);
-            require(consumed <= tank, "Gateway/not-enough-gas-funds");
-            tank -= consumed;
+                uint256 consumed = currentRouter.estimate(payload, isPrimaryRouter ? messageCost : proofCost);
 
-            currentRouter.pay{value: consumed}(payload, address(this));
-            currentRouter.send(payload);
-        }
+                require(consumed <= tank, "Gateway/not-enough-gas-funds");
+                tank -= consumed;
+
+                currentRouter.pay{value: consumed}(payload, address(this));
 
-        if (fuel > 0) quota = 0;
+                currentRouter.send(payload);
+            }
+            quota = 0;
+        } else if (gasService.shouldRefuel(source, message)) {
+            uint256 tank = address(this).balance;
+            for (uint256 i; i < numRouters; i++) {
```

Comment:
I know this loop is duplicated  but IMO the way it's structured like this makes it more readable and shows clearer intention. I'm happy to discuss it though. 
The body of the loop is not completely the same but the loop itself can be extracted in a separate function and just  pass some payment handling logic. I think it will get more complicated

Reply:
Yeah I went over this 3 times and came the same conclusion so far, it's not ideal but doesn't seem like any alternatives are cleaner :)